src/gui/inputdevices/TouchInputHandler.cpp

```diff
@@ -7,102 +7,175 @@
+	int vPadding = inputContext->getSettings()->getAddVerticalSpace() ? inputContext->getSettings()->getAddVerticalSpaceAmount() : 0;
+	int hPadding = inputContext->getSettings()->getAddHorizontalSpace() ? inputContext->getSettings()->getAddHorizontalSpaceAmount() : 0;
 
-		this->lastPosX = event->absoluteX;
-		this->lastPosY = event->absoluteY;
+	double centerX = (this->priLastRelX + this->secLastRelX) / 2.0 - hPadding;
+	double centerY = (this->priLastRelY + this->secLastRelY) / 2.0 - vPadding;
 
-		ZoomControl* zoomControl = this->inputContext->getView()->getControl()->getZoomControl();
-		std::tuple<double, double> pos = zoomControl->getScrollPositionAfterZoom();
+	this->startZoomDistance = std::pow(this->priLastAbsX - this->secLastAbsX, 2.0) + std::pow(this->priLastAbsY - this->secLastAbsY, 2.0);
+	lastZoomScrollCenterX = (this->priLastAbsX + this->secLastAbsX) / 2.0;
+	lastZoomScrollCenterY = (this->priLastAbsY + this->secLastAbsY) / 2.0;
 
-		double newX = std::get<0>(pos) - offsetX;
-		double newY = std::get<1>(pos) - offsetY;
+	ZoomControl* zoomControl = this->inputContext->getView()->getControl()->getZoomControl();
 
-		zoomControl->setScrollPositionAfterZoom(newX, newY);
+	// Disable zoom fit as we are zooming currently
+	// TODO this should happen internally!!!
+	if(zoomControl->isZoomFitMode())
+	{
+		zoomControl->setZoomFitMode(false);
 	}
 
-	//  Manually scroll if non-touchscreen device was mapped to a touchscreen (GTK wont handle this)
-	if (this->lastPosX >= 0.0 && this->lastPosY >= 0.0
-			&& event->deviceClass == INPUT_DEVICE_TOUCHSCREEN && gdk_device_get_source(gdk_event_get_source_device(event->sourceEvent)) != GDK_SOURCE_TOUCHSCREEN)
-	{
+	Rectangle zoomSequenceRectangle = zoomControl->getVisibleRect();
 
-		double offsetX = event->absoluteX - this->lastPosX;
-		double offsetY = event->absoluteY - this->lastPosY;
+	zoomControl->startZoomSequence(centerX - zoomSequenceRectangle.x, centerY - zoomSequenceRectangle.y);
+}
 
-		this->lastPosX = event->absoluteX;
-		this->lastPosY = event->absoluteY;
+void TouchInputHandler::zoomMotion(InputEvent* event)
+{
 
-		GtkAdjustment* h = this->inputContext->getView()->getScrollHandling()->getHorizontal();
-		gtk_adjustment_set_value(h, gtk_adjustment_get_value(h) - offsetX);
-		GtkAdjustment* v = this->inputContext->getView()->getScrollHandling()->getVertical();
-		gtk_adjustment_set_value(v, gtk_adjustment_get_value(v) - offsetY);
+	if (event->sequence == this->primarySequence)
+	{
+		this->priLastAbsX = event->absoluteX;
+		this->priLastAbsY = event->absoluteY;
+	} else {
+		this->secLastAbsX = event->absoluteX;
+		this->secLastAbsY = event->absoluteY;
 	}
-}
 
-void TouchInputHandler::actionEnd(InputEvent* event)
-{
-	XOJ_CHECK_TYPE(TouchInputHandler);
+	double sqDistance = std::pow(this->priLastAbsX - this->secLastAbsX, 2.0) + std::pow(this->priLastAbsY - this->secLastAbsY, 2.0);
+	double zoom = sqDistance / this->startZoomDistance;
```

Comment:
The real distances are std::sqrt(distance) for both of these values and I think that's whats causing the non-natural zoom. I suggest to change either this line to:
```
	double zoom = std::sqrt(sqDistance / this->startZoomDistance);
```
or store both the distances directly as _real_ distances.
(length of an vector = sqrt(x²+y²))

Reply:
Ohh I just copied this from your approach without looking at it^^ 
The new algorithm is better but still not perfect. I have absolutely no idea where the divergence might come from.

Reply:
It should be good enough with the new approach. It diverges by 5mm for 10cm movement which is more or less beneath the finger anyway.